### ISSUE TITLE
Combination needed for routing: "paved / asphalt" together with "bicycle = yes" 

### ISSUE THREAD

**Issue by Posteinwurf** [osmandapp/Osmand#3914]
Dear Team, 

I hope you are not getting tired of improving the bicycle routing. I would like to suggest another feature that will be a big step forward. And it might only take 10 minutes to program for someone who knows Java (I do not). 

Last week I had to work in Vienna, and so I could do some real-world-tests there. I am trying to further optimize the routing.xml so that it finds a really good way in ANY city. But there is still one problem: At the moment, the routing.xml can either read "bicycle=yes" OR "paved". The other value is ignored. What we need is a combination: A big upgrade for the ways that have bicycle=yes AND paved (and afterwards, a small upgrade for ways that ONLY have bicycle=yes, etc). 

I have now understood the commands for osmand_highway_integrity_brouting. A short java code like that could solve this problem easily! There are just a few steps that we would need:

1.) The following only applies to a) paths, b) tracks, c) footways, d) residential and e) pedestrian.

2.) create a new value called "bike_yes_and_paved". If you find "bicycle=yes" AND "asphalt" then set "bike_yes_and_paved" to 1. (Also set it to 1, if "bicycle=yes" is combined with paved or paving_stones or concrete or tracktype=grade1.)

3.) create a new value "bike_designated_and_paved" (again, set it to 1, if "bicycle=designated" is combined with any tag that indicates a paved surface).

This would very often lead to a better result in bicycle routing. Could we put this into MapRenderingTypesEncoder.java? That would be GREAT! 

Thank you in advance

Malte

**Comment by Posteinwurf** [comment]
Sorry to urge - this would be a big step forward for bicycle routing, and it would probably only take some minutes to implement. Could someone help me? I am working on some further improvements of routing.xml and I would need this as a basis.

Unfortunately, I am not able to do it myself (I could try, but maybe it would take less time for someone with Java knowledge to do it from Scratch).

**Comment by vshcherb** [comment]
@xmd5a2  should be able to help in 1-2 weeks. I hope Java knowledge will not be needed. 
This is what we used to preprocess OSM data
https://github.com/osmandapp/OsmAnd-resources/blob/master/obf_creation/rendering_types.xml

Please take a look at rules osm_convert and so on. You are going to the right direction with producing osmand_ artificial tags same way we did osmand_integrity coefficient.

**Comment by vshcherb** [comment]
rendering_types could be used custom with http://download.osmand.net/latest-night-build/OsmAndMapCreator-main.zip (check the settings please)

**Comment by Posteinwurf** [comment]
Thank you for the help! 1-2 weeks sounds good.

 

Do you want me to draft something? I could try, but it might be faster for you to do it from scratch! My knowledge is still limited on how to write a good code for rendering_types.xml.

**Comment by xmd5a2** [comment]
You can test code yourself using https://www.compilejava.net/
Here is my code for surface_integrity with test function https://pastebin.com/xDA6cfVv
Final code for this issue should be much more simpler.

**Comment by Posteinwurf** [comment]
Thank you! Is that the code that is currently being used for the surface integrity function in the bicycle routing?

 

I think it is too complicated for me – either I need some explanations or you suggest something for the paved / asphalt together with bicycle=yes.

**Comment by vshcherb** [comment]
The idea of using & combining tags sounds quite difficult:
bike_yes_and_paved, 
bike_yes_and_asphalt,
....
bike_no_...

Is not particularly great cause leads to too many combinations & makes routing.xml quite cumbersome. Do we plan to see advantages of it? 

It feels that bicycle=yes is an access tag, and paved/asphalt - priority tag, so it should not be a case to combine it. 

P.S. Vienna is a great city for bicycles and it has own cycleway/seggregated which should be prioritised today as well.

**Comment by Posteinwurf** [comment]
I have to explain further. My suggestion does not make routing more complicated AND it is a big step forward for bicycle routing. Let me explain:

**Not complicated:**
We only need TWO additional values for the routing.xml. And only two additional command lines in the routing.xml! So, it does not get complicated at all. The TWO additional values are:
1. **bicycle_designated_and_paved**. It is set to true, if a way is tagged bicycle=designated AND at the same time either "asphalt" or "paved" or "paving_stones" or "concrete" or "tracktype=grade1".
2. **bicycle_yes_and_paved.** Same as above, but  with the value "yes" instead of value "designated".

**Big step forward**
The two new values close a big gap that exists in the routing.xml at the moment. This gap cannot be closed by other means. Why is it so important?

The combination of bicycle=yes/designated with "paved" is ideal for bicycles. Often, these ways are even better than cycleways, because "bicycle=yes" often indicates ways through parks, etc, while cycleways often go along main roads.

But at the moment, the routing engine will EITHER read "bicycle=yes" and ignore "paved" OR it will read "paved" and ignore "bicycle=yes" (depending on the order of these values in the routing.xml). So at the moment (without the new values), we only have two very bad options. First option: If a way has bicycle=yes, then we completely ignore if it is paved or not. b) Second option: If a way is paved, then we completely ignore if it has the tag bicycle=yes. Both options are bad.

With the two new values, everything will be perfect:

First, we set a very high priority for all the ways that are paved and have bicycle=yes/designated

Now, the only ways that are still left, are the ways that have EITHER bicycle=yes/designated OR paved. So now, we can set a lower priority for these two values. But this lower priority will NOT apply to the ways that have already be given the high priority before.

**Comment by vshcherb** [comment]
I think we have something very similar already osmand_highway_integrity_brouting, so in that case it won't work together.

**Comment by Posteinwurf** [comment]
yes, the osmand_highway_integrity_brouting actually gave me the idea. So that was helpful!

So is there someone who could implement the two new combined values that we need?
